### PR TITLE
Introducing PHP8 compliancy

### PR DESCRIPTION
**Pull request by giovannialbero1992**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -23,7 +23,7 @@
     "doctrine/inflector": "^1.3",
     "mockery/mockery": "^1.2",
     "phpstan/phpstan": "^0.12",
-    "phpunit/phpunit": "^7.5",
+    "phpunit/phpunit": "^7.5|^9.3",
```
Comment (deguif) [inline composer.json:26]: Should be `||` as it's new standard ;)
See https://github.com/composer/composer/issues/6755#issuecomment-337588505

Comment (giovannialbero1992) [reply composer.json]: corrected :)

File: composer.json [hunk 2]
```diff
@@ -23,7 +23,7 @@
     "doctrine/inflector": "^1.3",
     "mockery/mockery": "^1.2",
     "phpstan/phpstan": "^0.12",
-    "phpunit/phpunit": "^7.5",
+    "phpunit/phpunit": "^7.5||^9.3",
```
Comment (OskarStark) [inline composer.json:26]: You should add spaces around the ||

Comment (giovannialbero1992) [reply composer.json]: done

File: .ci/packer_cache.sh [hunk 3]
```diff
@@ -2,10 +2,10 @@
 
 source /usr/local/bin/bash_standard_lib.sh
 
-DOCKER_IMAGES="php:7.4-cli
+DOCKER_IMAGES="8.0.0beta3-cli-buster
```
Comment (OskarStark) [inline .ci/packer_cache.sh:5]: AFAIK the first RC i out?! 🤔

File: .ci/packer_cache.sh [hunk 4]
```diff
@@ -2,10 +2,10 @@
...
+php:7.4-cli
 php:7.3-cli
 php:7.2-cli
-php:7.1-cli
```
Comment (OskarStark) [inline .ci/packer_cache.sh:8]: Is it not possible to keep 7.1 ?

File: composer.json [hunk 5]
```diff
@@ -23,7 +23,7 @@
     "doctrine/inflector": "^1.3",
     "mockery/mockery": "^1.2",
     "phpstan/phpstan": "^0.12",
-    "phpunit/phpunit": "^7.5",
+    "phpunit/phpunit": "^7.5 || ^9.3",
```
Comment (OskarStark) [inline composer.json:26]:
```suggestion
    "phpunit/phpunit": "^7.5 || ^9.3",
```

Why not supporting `8.x`, too?

File: tests/Elasticsearch/Tests/ClientBuilder/DummyLogger.php [hunk 6]
```diff
@@ -19,5 +19,5 @@
 
 class DummyLogger
 {
-    
+
```
Comment (OskarStark) [inline tests/Elasticsearch/Tests/ClientBuilder/DummyLogger.php:22]: Fixed in https://github.com/elastic/elasticsearch-php/pull/1068

Comment (giovannialbero1992) [reply tests/Elasticsearch/Tests/ClientBuilder/DummyLogger.php:22]: when your PR will be merged I'll resolve conflict.

**Comment by deguif** [comment]
@ezimuel Is there anything we can do to help moving this forward.

`ruflin/elastica` added support for PHP 8.0 recently but had to skip platform requirements while testing in its CI as `elasticsearch/elasticsearch` doesn't support PHP 8.0 in its composer requirement.

I think this should target branch `7.x`?

**Comment by mfn** [comment]
The title of the PR
> Introducing PHP8 compliancy

I'm starting to test run this library on PHP 8 in my company and found some issues (using v7.10.0) but was surprised this PR does not touch any of the code parts I saw errors 🤷‍♀️

For example, without a vendor ~patch~ hack I can't connect at all:
```php
ErrorException: Elasticsearch\Connections\Connection::Elasticsearch\Connections\{closure}(): Argument #2 ($key) must be passed by reference, value given

/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php:358
/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php:218
/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Transport.php:110
/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Client.php:1637
/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Client.php:310
…
```

In this case this is the diff for the "fix" (removed the unused ❗ `&$key` variable) in https://github.com/elastic/elasticsearch-php/blob/dfd2b62209cd0ec4f212e92bb5641b23caa667f1/src/Elasticsearch/Connections/Connection.php#L345 

```diff
Index: vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php b/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php
--- a/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php	(revision 8f1256488b3c58bec0e3027bbc4d74a2408204f0)
+++ b/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php	(revision 1c323624f0f741645ab9af0c66d6deffd01553e9)
@@ -349,7 +349,7 @@
         if (isset($params) === true && !empty($params)) {
             array_walk(
                 $params,
-                function (&$value, &$key) {
+                function (&$value) {
                     if ($value === true) {
                         $value = 'true';
                     } elseif ($value === false) {
```

Anyone else seeing this? Should I create an issue or PR directly? I'm just not sure as it seems these tests here seem to be green on PHP 8 nevertheless.

I'm also running tests on GHA using setup-php but I did run them recently, so maybe it's a new regression (or it's just simply not covered in the suite 🤷‍♀️)

Thanks

**Comment by deguif** [comment]
@mfn good catch.
See https://3v4l.org/ZmrdO to confirm that the code underlined by @mfn is breaking in PHP 8.

**Comment by mfn** [comment]
Smart 😄 Thank you @deguif 🙇

**Comment by deguif** [comment]
@mfn this should also be fixed by https://github.com/elastic/elasticsearch-php/pull/1075

**Comment by mfn** [comment]
@deguif Thank you, I confirmed it https://github.com/elastic/elasticsearch-php/pull/1075#issuecomment-728130942 !

**Comment by ezimuel** [comment]
@giovannialbero1992, @deguif, @mfn, @OskarStark thanks for your contribution! I'll update soon, thanks!

**Comment by pascaldevink** [comment]
@ezimuel @deguif PHP 8 was released today, is there any update on the progress of this PR

**Comment by reedy** [comment]
I guess with 8.0 being table, we shouldn't need to use RC4 anymore (I'm presuming newer images exist)...

**Comment by giovannialbero1992** [comment]
Hi @reedy, I updated the image used :smiley:

**Comment by ezimuel** [comment]
Hi @giovannialbero1992 sorry for the delay, I finally have some time to review the PR. I noticed that you sent it to `master`, can you rebase and sent to `7.x` branch? Thanks!

**Comment by ezimuel** [comment]
@giovannialbero1992 I fixed some PHP 8 compatibility issues:
- added version 3 of `cpliakas/git-wrapper` in composer;
- removed the annotation for expecting exceptions in `ClientBuilderTest.php`, for PHPUnit 9;
Moreover, I reintroduced the support of PHP 7.1, we still need to support it :scream: until next major of `elasticsearch-php`.

**Comment by ezimuel** [comment]
Welcome PHP 8!!!

**Comment by mfn** [comment]
Awesome!

FTR, I won't be really working with PHP8 until https://github.com/elastic/elasticsearch-php/pull/1075 is in, too!

**Pull request merged**